compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -1210,49 +1211,96 @@ impl<'local, 'interner> Interpreter<'local, 'interner> {
-                value => {
-                    let location = this.elaborator.interner.expr_location(&expr);
-                    let typ = value.get_type().into_owned();
-                    Err(InterpreterError::NonIntegerUsedInLoop { typ, location })
-                }
+        let start_value = self.evaluate(for_.start_range)?;
+        let signed = match start_value {
+            Value::I8(_) | Value::I16(_) | Value::I32(_) | Value::I64(_) => true,
+            Value::U1(_)
+            | Value::U8(_)
+            | Value::U16(_)
+            | Value::U32(_)
+            | Value::U64(_)
+            | Value::U128(_) => false,
+            value => {
+                let location = self.elaborator.interner.expr_location(&for_.start_range);
+                let typ = value.get_type().into_owned();
+                return Err(InterpreterError::NonIntegerUsedInLoop { typ, location });
             }
         };
 
-        let (start, make_value) = get_index(self, for_.start_range)?;
-        let (end, _) = get_index(self, for_.end_range)?;
         let was_in_loop = std::mem::replace(&mut self.in_loop, true);
 
         let mut result = Ok(Value::Unit);
 
-        for i in start..end {
-            self.push_scope();
-            self.current_scope_mut().insert(for_.identifier.id, make_value(i));
+        if signed {
+            // i128 can store all values from i8 - u64
+            let get_index = |this: &mut Self, expr| -> IResult<(_, fn(_) -> _)> {
+                match this.evaluate(expr)? {
+                    Value::I8(value) => Ok((value as i128, |i| Value::I8(i as i8))),
+                    Value::I16(value) => Ok((value as i128, |i| Value::I16(i as i16))),
+                    Value::I32(value) => Ok((value as i128, |i| Value::I32(i as i32))),
+                    Value::I64(value) => Ok((value as i128, |i| Value::I64(i as i64))),
+                    value => unreachable!("Checked above that value is signed type"),
+                }
+            };
 
-            let must_break = match self.evaluate(for_.block) {
-                Ok(_) => false,
-                Err(InterpreterError::Break) => true,
-                Err(InterpreterError::Continue) => false,
-                Err(error) => {
-                    result = Err(error);
-                    true
+            let (start, make_value) = get_index(self, for_.start_range)?;
+            let (end, _) = get_index(self, for_.end_range)?;
+
+            for i in start..end {
+                self.push_scope();
+                self.current_scope_mut().insert(for_.identifier.id, make_value(i));
+
+                let must_break = match self.evaluate(for_.block) {
+                    Ok(_) => false,
+                    Err(InterpreterError::Break) => true,
+                    Err(InterpreterError::Continue) => false,
+                    Err(error) => {
+                        result = Err(error);
+                        true
+                    }
+                };
```

Comment:
It's not a lot of code duplication but we could have a helper function like this:

```rust
fn evaluate_loop_body(&mut self, body: ExprId, result: &mut IResult<Value>) -> bool {
    match self.evaluate(body) {
        Ok(_) => false,
        Err(InterpreterError::Break) => true,
        Err(InterpreterError::Continue) => false,
        Err(error) => {
            *result = Err(error);
            true
        }
    }
}
```

then call it like this:

```rust
let must_break = self.evaluate_loop_body(for_.block, &mut result);
```

And we could also use this for `loop` and `while` handling.